provisioning/transport/http/src/http.ts

```diff
@@ -0,0 +1,251 @@
+            callback(new errors.InvalidOperationError('registration already in progress.'));
+          },
+          disconnect : (callback) => {
+            this._cancelCurrentOperation((err) => {
+              this._fsm.transition('idle', callback, err);
+            });
+          }
+        },
+        polling: {
+          _onEnter: (callback) => {
+            this._sendOperationStatusRequest((err, body) => {
+              if (this._registrationInProgress()) { // make sure we weren't cancelled before doing something with the response
+                if (err) {
+                  this._fsm.transition('idle', callback, err);
+                } else if (body.status.toLowerCase() === 'assigning') {
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_028: [ If the operation status response contains a success status code with a 'status' that is 'Assigning', `register` shall fire an `operationStatus` event with the deserialized body and continue polling. ] */
+                  this.emit('operationStatus', body);
+                  this._fsm.transition('waiting_to_poll', callback);
+                } else if (body.status.toLowerCase() === 'assigned') {
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_027: [ If the operation status response contains a success status code with a 'status' of 'Assigned', `register` shall stop polling and call the `callback` with `err` == null and the body containing the deserialized body. ] */
+                  this._fsm.transition('idle', callback, null, body);
+                } else {
+                  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_030: [ If the operation status response has a success code with a 'status' that is any other value, `register` shall call the callback with a `SyntaxError` error and stop polling. ] */
+                  this._fsm.transition('idle', callback, new SyntaxError('badly formed registration response'));
+                }
+              }
+            });
+          },
+          register: (authorization, forceRegistration, body, callback) => {
+            /* Codes_SRS_NODE_PROVISIONING_HTTP_18_036: [ `register` shall call the `callback` with an `InvalidOperationError` if it is called while a previous registration is in progress. ] */
+            callback(new errors.InvalidOperationError('registration already in progress.'));
+          },
+          disconnect : (callback) => {
+            this._cancelCurrentOperation((err) => {
+              this._fsm.transition('idle', callback, new errors.OperationCancelledError());
+            });
+          }
+        }
+      }
+    });
+  }
+
+  register(authorization: string | X509, forceRegistration: boolean, body: any, callback: ProvisioningResponseCallback): void {
+    this._fsm.handle('register', authorization, forceRegistration, body, callback);
+  }
+
+  disconnect(callback: (err: Error) => void): void {
```

Comment:
>disconnect [](start = 2, length = 10)

disconnect is kindof weird for a REST based thing. since we're using cancel in other places, wouldn't it be logical to call it cancel? or cancelCurrentOperation?

also, do other SDKs support cancelling a registration process? 

last but not least will we be able to do that with HTTP?

also, does the service cares to know we're cancelling?

Reply:
I agree.  I will sort this out after I think about the other protocols for a while.  The function name is a placeholder until I get a unified interface

---
In reply to: [142299914](https://github.com/Azure/azure-iot-sdk-node/pull/99#discussion_r142299914) [](ancestors = 142299914)

Reply:
please make sure for all TODO list items that you have associated PBIs. or else we're going to miss things.

---
In reply to: [142460536](https://github.com/Azure/azure-iot-sdk-node/pull/99#discussion_r142460536) [](ancestors = 142460536,142299914)